### ISSUE TITLE
Mixing parameterized role into enum value causes compile-time error

### ISSUE THREAD

**Issue by landyacht** [rakudo/rakudo#5515]
## The Problem

With more than one value in an `enum`, mixing a parameterized role into the value yields the obtuse comile-time error "Incompatible MROs in P6opaque rebless for types Int+{MixedInRoleName[Str]} mixin and EnumName". This does not happen when there is only one enum value, nor when the role is unparameterized. 

## Steps to Reproduce
```raku
role MixedInRoleName[$a] { has $.a = $a; }
enum EnumName (
    KeyOne => 1 but MixedInRoleName['Yksi'],
    KeyTwo => 2 but MixedInRoleName['Kaksi']
);
```

## Environment
* Operating system: Linux 6.1.53-gentoo-r1 (reproduced also with raku eval bot in IRC)
* Raku version: Rakudo 2023.12 / MoarVM 2023.12

**Comment by lizmat** [comment]
ping @vrurg

**Comment by vrurg** [comment]
It lurks in my memory that values of an enum has to be of the same type. No way for me to recall where I saw this, neither I can find the place at the moment.

Trying to mix a number and a string results in exactly the same error. Apparently, the error itself is LTA, but otherwise it seems to be legal.

The particular example case can be easily fixed with this:

```raku
role MixedInRoleName { has $.a; }
enum EnumName (
    KeyOne => (1 but MixedInRoleName('Yksi')),
    KeyTwo => (2 but MixedInRoleName('Kaksi'))
);
say KeyOne.a;
```

_Minor offtop:_ I still remember myself thinking "what the hell this 'mixin attribute' means and why would I need it?" :)

**Comment by landyacht** [comment]
Interesting. I suppose technically roles that are parameterized differently do cause the resulting type to be different. Still, might be worth leaving this open to improve the error message?

**Comment by vrurg** [comment]
> Still, might be worth leaving this open to improve the error message?

Absolutely.

**Comment by vrurg** [comment]
> I suppose technically roles that are parameterized differently do cause the resulting type to be different.

Yes, you're correct. This is why the approach you took in your example is not the best one in some scenarios. Imagine one having thousands of objects to be marked with roles-labels. In this case they get thousands of additional types, each consuming memory both to hold its VM representation and to maintain metaobject data related to the type.

**Comment by landyacht** [comment]
Following up on this, perhaps it makes sense to take the closest common ancestor of the types provided by the user, or if that's too obtuse given multiple inheritance, allow the user to specify the value type. Being able to mix roles into an enum like this would be nice.
